src/lowering/desugar.ml

```diff
@@ -658,16 +658,10 @@ and text_dotE proj e =
     |  _ -> assert false
 
 and let_else_switch p e f =
-  {
-    e with
-    it = S.SwitchE(
-      e,
-      S.[
-        { it = { pat = p; exp = e }; at = e.at; note = () };
-        { it = { pat = { it = WildP; at = f.at; note = p.note }; exp = f }; at = f.at ; note = () }
-      ]
-    )
-  }
+  exp { e with it = S.SwitchE(e,
+        S.[ { it = { pat = p; exp = e }; at = e.at; note = () };
```

Comment:
This has a small problem. If there are side effects in `e`, then those will happen potentially twice, since `e` is not used linearly. So theoretically we have to bind e to a fresh binding before doing the `SwitchE` and use that twice.

Reply:
Maybe do a `debugPrint`-test first to see the problem and then we'll have to fix this sometime.

Reply:
Btw. `and` patterns would solve this neatly:
``` Motoko
let <pat> = switch <exp> {
    case (<pat> and binding) binding;
    case _ fail
}
```
Alas, we don't have them (yet).

Reply:
Ah! Yes, side effects. Didn't think of that. Will take a look

Reply:
So currently:
``` 
     func () { 
       var y = 0; 
       let x = (y := 2) else { return }; 
     };
```
compiles to
```(VarD y Nat (LitE (NatLit 0)))
        (LetD
          (VarP x)
          (SwitchE
            (AssignE (VarLE y) (LitE (NatLit 2)))
            (case (VarP x) (AssignE (VarLE y) (LitE (NatLit 2))))
            (case WildP (PrimE RetPrim (PrimE TupPrim)))
          )
        )```

which is going to evaluate the assignment to `y` twice

Reply:
Uhh bad idempotent example, but you get the point haha

Reply:
Okay this should be fixed now

Reply:
btw., easy non-idempotents are `v += 1` etc.